subsys/net/lib/ptp/transport.c

```diff
@@ -0,0 +1,354 @@
+		LOG_ERR("Failed to set ip multicast ttl socket option");
+		goto error;
+	}
+
+	if (zsock_getsockopt(socket, IPPROTO_IP, IP_TOS, &tos, &length)) {
+		tos = 0;
+	}
+
+	tos &= ~0xFC;
+	tos = CONFIG_PTP_DSCP_VALUE << 2;
+	length = sizeof(tos);
+
+	if (zsock_setsockopt(socket, IPPROTO_IP, IP_TOS, &tos, length)) {
+		LOG_WRN("Failed to set DSCP priority");
+	}
+
+	return socket;
+error:
+	zsock_close(socket);
+	return -1;
+}
+
+static int transport_udp_ipv6_open(struct net_if *iface, uint16_t port)
+{
+	uint8_t tclass;
+	socklen_t length;
+	int socket, hops = 1, feature_on = 1;
+	struct sockaddr_in6 addr = {
+		.sin6_family = AF_INET6,
+		.sin6_addr = IN6ADDR_ANY_INIT,
+		.sin6_port = htons(port)
+	};
+
+	socket = transport_socket_open(iface, (struct sockaddr *)&addr);
+
+	if (socket < 0) {
+		return -1;
+	}
+
+	if (zsock_setsockopt(socket,
+			     IPPROTO_IPV6,
+			     IPV6_RECVPKTINFO,
+			     &feature_on,
+			     sizeof(feature_on))) {
+		LOG_ERR("Failed to set IPV6_RECVPKTINFO");
+		goto error;
+	}
+
+	if (zsock_setsockopt(socket, IPPROTO_IPV6, IPV6_MULTICAST_HOPS, &hops, sizeof(hops))) {
+		LOG_ERR("Failed to set ip multicast hops socket option");
+		goto error;
+	}
+
+	if (zsock_getsockopt(socket, IPPROTO_IPV6, IPV6_TCLASS, &tclass, &length)) {
+		tclass = 0;
+	}
+
+	tclass &= ~0xFC;
+	tclass = CONFIG_PTP_DSCP_VALUE << 2;
+	length = sizeof(tclass);
+
+	if (zsock_setsockopt(socket, IPPROTO_IPV6, IPV6_TCLASS, &tclass, length)) {
+		LOG_WRN("Failed to set priority");
+	}
+
+	return socket;
+error:
+	zsock_close(socket);
+	return -1;
+}
+
+static int transport_send(int socket, int port, void *buf, int length, struct sockaddr *addr)
+{
+	struct sockaddr m_addr;
+	socklen_t addrlen;
+	int cnt;
+
+	if (!addr) {
+		if (IS_ENABLED(CONFIG_PTP_UDP_IPv4_PROTOCOL)) {
+			m_addr.sa_family = AF_INET;
+			net_sin(&m_addr)->sin_port = htons(port);
+			net_sin(&m_addr)->sin_addr.s_addr = mcast_addr.s_addr;
+
+		} else if (IS_ENABLED(CONFIG_PTP_UDP_IPv6_PROTOCOL)) {
+			m_addr.sa_family = AF_INET6;
+			net_sin6(&m_addr)->sin6_port = htons(port);
+			memcpy(&net_sin6(&m_addr)->sin6_addr,
+			       &mcast_addr,
+			       sizeof(struct in6_addr));
+		}
+		addr = &m_addr;
+	}
+
+	addrlen = IS_ENABLED(CONFIG_PTP_UDP_IPv4_PROTOCOL) ? sizeof(struct sockaddr_in) :
+							     sizeof(struct sockaddr_in6);
+	cnt = zsock_sendto(socket, buf, length, 0, addr, addrlen);
+
+	if (cnt < 1) {
+		LOG_ERR("Failed to send message");
+		return -EFAULT;
+	}
+
+	return cnt;
+}
+
+int ptp_transport_open(struct ptp_port *port)
+{
+	static const int socket_ports[] = {PTP_SOCKET_PORT_EVENT, PTP_SOCKET_PORT_GENERAL};
+	int socket;
+
+	if (net_addr_pton(IS_ENABLED(CONFIG_PTP_UDP_IPv4_PROTOCOL) ? AF_INET : AF_INET6,
+		IS_ENABLED(CONFIG_PTP_UDP_IPv4_PROTOCOL) ? IP_MULTICAST_IP : IP6_MULTICAST_IP,
+		&mcast_addr)) {
+		LOG_ERR("Couldn't resolve multicast IP address");
+		return -1;
+	}
+
+	for (int i = 0; i < PTP_SOCKET_CNT; i++) {
+		socket = IS_ENABLED(CONFIG_PTP_UDP_IPv4_PROTOCOL) ?
```

Comment:
There was also `PTP_IEEE_802_3_PROTOCOL` choice option, what about it? Currently it seems ignored/not used, maybe it should be removed for now to avoid confusion.

Reply:
I didn't implement that as there is gPTP that uses that and that would increase the PR size that is already big. Sure I'll remove it and add the option when submitting support for the protocol.